Initial Implementation of Hydrator++ using Redux

ajainarayanan opened the pull request:
- Adds `HydratorPlusPlusLeftPanelStore` as a `Redux` store.
- Adds `HydratorPlusPlusActions` `action creator` to interact with `Redux` store asynchronously.
- Adds `ReduxThunk` for asyn communication with the store.
- Adds unit test for `HydratorPlusPlusLeftPanelStore`.
- Removes unstable (old) unit tests for `mySettings` & `mySocket` services.
##### TODO:
- Need to write unit tests for `HydratorPlusPlusActions` service
- Need to write more negative unit tests for `HydratorPlusPlusLeftPanelStore` service.

ajainarayanan committed:
Adds Redux, Redux-Thunk & used in rewriting left panel store in Hydrator++.

The left panel store is re-written to use Redux to be implemented in a more functional way.
This way we have a more functional approach with unit tests.

This commit rewrites LeftPanelStore as a map of plugin types and list of extensions. This is exactly how we show
things in left panel. Extensions include "batchsource", "batchsink" etc., and plugin types map
maintains a map of each extension to plugins (mapping each plugin with a list of all artifact versions supported by backend).

elfenheart commented on cdap-ui/test/unit/settings-test.js:
```diff
@@ -1,68 +1,68 @@
\ No newline at end of file
+// /*
+//  * Copyright © 2015 Cask Data, Inc.
+//  *
```
just remove the file?

ajainarayanan replied:
No. We still need to fix this test. `mySettings` could be tested. It should be simple validations.

elfenheart commented:
Getting this error:

![screen shot 2016-05-04 at 11 42 42 am](https://cloud.githubusercontent.com/assets/4398643/15025288/63f481e4-11ed-11e6-86ff-4ee2d3f752ad.png)

ajainarayanan commented:
~~@elfenheart Did you do `npm install`? https://github.com/caskdata/cdap/blob/feature/ui-redux-hydrator%2B%2B/cdap-ui/package.json#L60~~
Alright updated the npm-shrinkwrap.json file.

elfenheart commented:
@ajainarayanan yup, I did a cache clean and do a fresh install as well.

elfenheart commented:
It seems like there is an issue when there is a state change while you are in studio. Getting JS error. To reproduce, try out changing Pipeline Type

ajainarayanan commented:
@elfenheart Verified the issues you mentioned and fixed. Added nodes to the canvas. Created a pipeline. Published it and ran it. Switched pipelines. Imported a pipeline. Chose a template. Went through possible scenarios I could think of. Could you please take another pass?

elfenheart commented:
On hold for plugin template display name / filtering

elfenheart commented:
👍  LGTM
